File: AndroidJavaDemo/app/src/main/java/com/example/androidjavademo/FirstFragment.java
package com.example.androidjavademo;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.NavHostFragment;

import com.example.androidjavademo.databinding.FragmentFirstBinding;

import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.StatusCode;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.context.Scope;

public class FirstFragment extends Fragment {

    private FragmentFirstBinding binding;

    @Override
    public View onCreateView(
            LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState
    ) {

        binding = FragmentFirstBinding.inflate(inflater, container, false);
        return binding.getRoot();

    }

    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        binding.buttonFirst.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // 获取Tracer
                Tracer tracer = OpenTelemetryUtil.getTracer();
                // 创建Span
                Span span = tracer.spanBuilder("First Fragment Button onClick").startSpan();
                try (Scope scope = span.makeCurrent()) {
                    // 获取traceId
                    System.out.println(span.getSpanContext().getTraceId());
                    // 获取spanId
                    System.out.println(span.getSpanContext().getSpanId());
                    // 设置属性
                    span.setAttribute("key", "value");

                    Attributes eventAttributes = Attributes.of(
                            AttributeKey.stringKey("key"), "value",
                            AttributeKey.longKey("result"), 0L);

                    // 添加事件
                    span.addEvent("onClick", eventAttributes);

                    parentSpan();

                    NavHostFragment.findNavController(FirstFragment.this)
                            .navigate(R.id.action_FirstFragment_to_SecondFragment);
                } catch (Throwable t) {
                    span.setStatus(StatusCode.ERROR, "Something wrong in onClick");
                    throw t;
                } finally {
                    span.end();
                }
            }
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    public void parentSpan() {
        // 获取Tracer
        Tracer tracer = OpenTelemetryUtil.getTracer();
        // 创建Span
        Span span = tracer.spanBuilder("Parent Span").startSpan();
        try (Scope scope = span.makeCurrent()) {
            // 获取traceId
            System.out.println(span.getSpanContext().getTraceId());
            // 获取spanId
            System.out.println(span.getSpanContext().getSpanId());
            childSpan();
        } finally {
            span.end();
        }
    }

    public void childSpan() {
        // 获取Tracer
        Tracer tracer = OpenTelemetryUtil.getTracer();
        // 创建Span
        Span span = tracer.spanBuilder("Child Span").startSpan();
        try (Scope scope = span.makeCurrent()) {
            // 获取traceId
            System.out.println(span.getSpanContext().getTraceId());
            // 获取spanId
            System.out.println(span.getSpanContext().getSpanId());
        } finally {
            span.end();
        }
    }

}

File: AndroidKotlinDemo/app/src/main/java/com/example/androidkotlindemo/FirstFragment.kt
package com.example.androidkotlindemo

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import com.example.androidkotlindemo.databinding.FragmentFirstBinding
import io.opentelemetry.api.common.AttributeKey
import io.opentelemetry.api.common.Attributes
import io.opentelemetry.api.trace.StatusCode
import io.opentelemetry.api.trace.Tracer

/**
 * A simple [Fragment] subclass as the default destination in the navigation.
 */
class FirstFragment : Fragment() {

    private var _binding: FragmentFirstBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {

        _binding = FragmentFirstBinding.inflate(inflater, container, false)
        return binding.root

    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.buttonFirst.setOnClickListener {
            // 获取Tracer
            val tracer: Tracer = OpenTelemetryUtil.getTracer()!!
            // 创建Span
            val span = tracer.spanBuilder("First Fragment Button onClick").startSpan()
            try {
                span.makeCurrent().use { scope ->
                    // 获取traceId
                    println(span.spanContext.traceId)
                    // 获取spanId
                    println(span.spanContext.spanId)
                    // 设置属性
                    span.setAttribute("key", "value")
                    val eventAttributes =
                        Attributes.of(
                            AttributeKey.stringKey("key"), "value",
                            AttributeKey.longKey("result"), 0L
                        )

                    // 添加事件
                    span.addEvent("onClick", eventAttributes)
                    findNavController().navigate(R.id.action_FirstFragment_to_SecondFragment)
                    parentSpan()
                }
            } catch (t: Throwable) {
                // 为Span设置状态
                span.setStatus(StatusCode.ERROR, "Something wrong in onClick")
                throw t
            } finally {
                span.end()
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    // 嵌套Span
    fun parentSpan() {
        // 获取Tracer
        val tracer: Tracer = OpenTelemetryUtil.getTracer()!!
        // 创建Span
        val span = tracer.spanBuilder("Parent Span").startSpan()
        try {
            span.makeCurrent().use { scope ->
                // 获取traceId
                println(span.spanContext.traceId)
                // 获取spanId
                println(span.spanContext.spanId)
                childSpan()
            }
        } finally {
            span.end()
        }
    }

    // 嵌套Span
    fun childSpan() {
        // 获取Tracer
        val tracer: Tracer = OpenTelemetryUtil.getTracer()!!
        // 创建Span
        val span = tracer.spanBuilder("Child Span").startSpan()
        try {
            span.makeCurrent().use { scope ->
                // 获取traceId
                println(span.spanContext.traceId)
                // 获取spanId
                println(span.spanContext.spanId)
            }
        } finally {
            span.end()
        }
    }
}

